Document/0x05c-Reverse-Engineering-and-Tampering.md

```diff
@@ -1272,6 +1272,85 @@ Now you can run the validate binary in your Android device to verify the solutio
+
+<android.content.Intent: android.content.Intent putExtra(java.lang.String,java.lang.CharSequence)> -> _SINK_
+<android.content.Intent: android.content.Intent putExtra(java.lang.String,char)> -> _SINK_
+<android.content.Intent: android.content.Intent putExtra(java.lang.String,java.lang.String)> -> _SINK_
+```
+
+To invoke FlowDroid via command line, following command can be used:
+
+```shell
+java -jar soot-infoflow-cmd/target/soot-infoflow-cmd-jar-with-dependencies.jar \
+    -a InsecureShop.apk \
+    -p Android/Sdk/platforms \
+    -s source_sink.txt
+```
+
+The output from the above command is following:
+
+```
+...
+
+[main] INFO soot.jimple.infoflow.android.SetupApplication$InPlaceInfoflow - The sink virtualinvoke r2.<android.content.Intent: android.content.Intent putExtra(java.lang.String,java.lang.String)>("password", $r5) in method <com.insecureshop.AboutUsActivity: void onSendData(android.view.View)> was called with values from the following sources:
+
+[main] INFO soot.jimple.infoflow.android.SetupApplication$InPlaceInfoflow - - $r1 = interfaceinvoke $r2.<android.content.SharedPreferences: java.lang.String getString(java.lang.String,java.lang.String)>("password", "") in method <com.insecureshop.util.Prefs: java.lang.String getPassword()>
+
+...
+
+[main] INFO soot.jimple.infoflow.android.SetupApplication$InPlaceInfoflow - The sink virtualinvoke r2.<android.content.Intent: android.content.Intent putExtra(java.lang.String,java.lang.String)>("username", $r4) in method <com.insecureshop.AboutUsActivity: void onSendData(android.view.View)> was called with values from the following sources:
+
+[main] INFO soot.jimple.infoflow.android.SetupApplication$InPlaceInfoflow - - $r1 = interfaceinvoke $r2.<android.content.SharedPreferences: java.lang.String getString(java.lang.String,java.lang.String)>("username", "") in method <com.insecureshop.util.Prefs: java.lang.String getUsername()>
+
+...
+
+[main] INFO soot.jimple.infoflow.android.SetupApplication - Found 2 leaks
+```
+
+The output indicates there are 2 leaks in the application, with one each corresponding to the username and password.
+
+The above output uses [jimple intermediate representation](https://www.sable.mcgill.ca/soot/doc/soot/jimple/Jimple.html "Jimple") to show the sources and sinks. Jimple representation contains enough information to locate the sources and sinks in the actual application. Since, InsecureShop app is an open source app, we can look for the source code corresponding to above findings, and shown below:
+
+```java
+// file: AboutActivity.kt
+
+fun onSendData(view: View) {
+        val userName = Prefs.username!!
+        val password = Prefs.password!!
+
+        val intent = Intent("com.insecureshop.action.BROADCAST")
+        intent.putExtra("username", userName)
+        intent.putExtra("password", password)
+        sendBroadcast(intent)
+
+        textView.text = "InsecureShop is an intentionally designed vulnerable android app built in Kotlin."
+
+    }
+```
```

Comment:
Consider adding a `blank` line after this code block

Reply:
done